File: dsp04.c
#include <graph.h>
#include <conio.h>
#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define HDIM 120
#define XDIM 120

float   fH[HDIM + 1], fX[XDIM + 1];

void main (void)
    {
    /*
     ***** PROGRAM NO.4  IMPULSE RESPONSE OF LTI PROCESSOR *****
     */
    int    iCount;
    float  fA1, fA2, fA3, fB0, fB1, fB2, fMax;

    _clearscreen  (_GCLEARSCREEN);
    printf ("PROGRAM 4: Impulse response of LTI processor\n");

    /*
     ****************** LOAD IMPULSE INTO INPUT ARRAY **********
     */
    printf ("Enter 3 recursive multipliers :");
    scanf ("%f, %f, %f", &fA1, &fA2, &fA3);

    printf ("\nEnter 3 nonrecursive multipliers :");
    scanf ("%f, %f, %f", &fB0, &fB1, &fB2);

    fX[10] = 1.0;

    /*
     **** FIND IMPULSE RESPONSE, TAKING N=10 AS TIME ORIGIN ****
     */

    printf ("Press any key to start...\n\n");
    getch();

    for (iCount = 10; iCount <= HDIM; iCount++)
        fH[iCount] = (fA1 * fH[iCount-1] +
                      fA2 * fH[iCount-2] +
                      fA3 * fH[iCount-3] +
                      fB0 * fX[iCount] +
                      fB1 * fX[iCount-1] +
                      fB2 * fX[iCount-2]);

    /*
     **** NORMALISE fH[iCount] TO A MAXIMUM VALUE OF UNITY *****
     */
    fMax = 0.0;
    for (iCount = 1; iCount <= HDIM; iCount++)
        if (fabs (fH[iCount]) > fMax)
            fMax = fabs (fH[iCount]);

    for (iCount = 1; iCount <= HDIM; iCount++)
        fH[iCount] /= fMax;

    _setvideomode( _MRES16COLOR );

    /*
     ********* PLOT HORIZONTAL AXIS AND IMPULSE INPUT **********
     */
    _moveto (0, 50);
    _lineto (300, 50);

    for (iCount = 1; iCount <= 100; iCount++)
        {
        _moveto (iCount * 3, 50);
        _lineto (iCount * 3, (int)(50.0 - fX[iCount+9] * 25.0 + 0.5));
        }

    /*
     ******* PLOT HORIZONTAL AXIS AND IMPULSE RESPONSE *********
     */
    _moveto (0, 120);
    _lineto (300, 120);

    for (iCount = 1; iCount <= 100; iCount++)
        {
        _moveto (iCount * 3, 120);
        _lineto (iCount * 3, (int)(120 - (fH[iCount+9] * 60.0) + 0.5));
        }

    getch();
    _setvideomode( _DEFAULTMODE );

    /*
     ***********************************************************
     */
    }

File: dsp04.pas
(******  PROGRAM NO.4  IMPULSE RESPONSE OF LTI PROCESSOR  ******)
program DSP04;
uses graph3;
var H: array[1..120] of real;
var X: array[1..120] of real;
var N: integer;
var A1,A2,A3,B0,B1,B2,M: real;
begin
(***************  LOAD IMPULSE INTO INPUT ARRAY  ***************)
  for N:=1 to 120 do
    begin
      X[N]:=0;H[N]:=0
    end;
  X[10]:=1.0;
  writeln('ENTER 3 RECURSIVE MULTIPLIERS');
  readln(A1,A2,A3);
  writeln('ENTER 3 NONRECURSIVE MULTIPLIERS');
  readln(B0,B1,B2);
(*****  FIND IMPULSE RESPONSE, TAKING N=10 AS TIME ORIGIN ******)
  for N:=10 TO 120 do
    begin
      H[N]:= A1*H[N-1]+A2*H[N-2]+A3*H[N-3]+B0*X[N]+B1*X[N-1]
                                                      +B2*X[N-2];
    end;
(*********  NORMALISE H[N] TO A MAXIMUM VALUE OF UNITY  ********)
  M:=0;
  for N:=1 to 120 do
    begin
      if abs(H[N])>M then M:=abs(H[N]);
    end;
  for N:=1 TO 120 do
    begin
      H[N]:=H[N]/M;
    end;
(***********  PLOT HORIZONTAL AXIS AND IMPULSE INPUT  **********)
  graphmode;
  draw(0,50,300,50,1);
  for N:=1 to 100 do
    begin
      draw(3*N,50,3*N,round(50-X[N+9]*25),1);
    end;
(*********  PLOT HORIZONTAL AXIS AND IMPULSE RESPONSE  *********)
  draw (0,120,300,120,1);
  for N:=1 to 100 do
    begin
      draw(3*N,120,3*N,round(120-H[N+9]*60),1);
    end;
end.
(***************************************************************)